src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,256 @@
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> repoLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            updateIfMatchesOneLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Update current state if there is exactly one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private void updateIfMatchesOneLabel(String keyword) {
+        if (TurboLabel.hasExactlyOneMatchedLabel(repoLabels, keyword)) {
+            updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(repoLabels, keyword));
+        }
+    }
+
+    /**
+     * Updates assignedLabels based on membership of a label i.e belonging to
+     * initialLabels, removedLabels or addedLabels.
+     * AssignedLabels are inferred from the updated addedLabels and removedLabels
+     * @param label
+     */
+    public void updateAssignedLabels(TurboLabel label) {
+        String labelName = label.getFullName();
+
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                removeConflictingLabels(label);
+                removedLabels.remove(labelName);
+            } else {
+                removedLabels.add(labelName);
+            }
+        } else {
+            if (isAnAddedLabel(labelName)) {
+                addedLabels.remove(labelName);
+            } else {
+                // add new label
+                removeConflictingLabels(label);
+                addedLabels.add(labelName);
+            }
+        }
+    }
+
+    /**
+     * Updates the list of labels which labels' names contain the current query.
+     * This list of labels can then be retrieved later via getMatchedLabels()
+     *
+     * The suggestion index will be pointed to the first label that fits the query, if there is.
+     *
+     * @param query
+     */
+    private void updateMatchedLabels(String query) {
```

Comment:
Is `query` a single word? In that case, we should call it `keyword` (to be consistent with the previous method, and also to make it clear that it is a single word)?